lib/smack/DSAWrapper.cpp

```diff
@@ -214,24 +102,104 @@ unsigned DSAWrapper::getPointedTypeSize(const Value *v) {
     llvm_unreachable("Type should be pointer.");
 }
 
-int DSAWrapper::getOffset(const Value *v) {
-  return getOffset(new MemoryLocation(v));
+unsigned DSAWrapper::getOffset(const Value *v) {
+  if (!DG->hasCell(*v))
+    return 0;
+  return DG->getCell(*v).getOffset();
+}
+
+const sea_dsa::Node *DSAWrapper::getNode(const Value *v) {
+  // For sea-dsa, a node is obtained by getting the cell first.
+  // It's possible that a value doesn't have a cell, e.g., undef.
+  if (!DG->hasCell(*v))
+    return nullptr;
+  auto node = DG->getCell(*v).getNode();
+  assert(node && "Values should have nodes if they have cells.");
+  return node;
 }
 
-// TODO: Should this return the node or its leader?
-const DSNode *DSAWrapper::getNode(const Value *v) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  auto *N = nodeEqs->getMemberForValue(v);
-  return N ? eqs.getLeaderValue(N) : nullptr;
+bool DSAWrapper::isTypeSafe(const Value *v) {
+  typedef std::unordered_map<unsigned, bool> FieldMap;
+  typedef std::unordered_map<const sea_dsa::Node *, FieldMap> NodeMap;
+  static NodeMap nodeMap;
+
+  auto node = getNode(v);
+
+  if (node->isOffsetCollapsed() || node->isExternal() || node->isIncomplete() ||
```

Comment:
Variable `node` could be `nullptr` here, and so we need to check that first.

Reply:
Good catch! I think we should check nullptr first and then call this function depending on the result. Please see my latest commit.